### PR TITLE
[Testing] CoW Protocol Batches

### PR DESCRIPTION
**Pull request by bh2smith**
Testing the issue raised in #1713 

PoC Query introduced in unit test (does not currently return no results as expected):

https://dune.com/queries/1389623

### CODE REVIEW

File: tests/cow_protocol/ethereum/cow_protocol_ethereum_assert_batches.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+
+with batches_with_trade_count as (
+    select
+        s.evt_tx_hash,
+        s.evt_block_time,
+        count(t.evt_index) as num_trades
+    from gnosis_protocol_v2_ethereum.GPv2Settlement_evt_Trade t
```
Comment (dot2dotseurat) [inline tests/cow_protocol/ethereum/cow_protocol_ethereum_assert_batches.sql:7]: can you reference these tables as sources?

**Comment by dot2dotseurat** [comment]
(looking into why the ref isn't being generated)

**Comment by dot2dotseurat** [comment]
@bh2smith I edited the script to pick up the ref from a modifed test. It's creating the dependency now but the test itself seems to be failing.

**Comment by bh2smith** [comment]
The test is expected to fail because there is an issue with the data itself. This is why I added the test. There is more context in the issue linked in the description.

File: tests/cow_protocol/ethereum/cow_protocol_ethereum_assert_batches.sql (hunk 1 above)
Comment (bh2smith) [reply tests/cow_protocol/ethereum/cow_protocol_ethereum_assert_batches.sql:7]: Yes absolutely.

**Comment by bh2smith** [comment]
So this is one reason why there is a missing trade and a resulting missing batch:

https://github.com/duneanalytics/spellbook/issues/1765

**Comment by bh2smith** [comment]
Now that #1765 has been fixed, this table is still showing thousands of missing records. However, the last one appeared 2 days ago. Could someone please rebuild this table and see if its really fixed? cc @dot2dotseurat

**Comment by dot2dotseurat** [comment]
> Now that #1765 has been fixed, this table is still showing thousands of missing records. However, the last one appeared 2 days ago. Could someone please rebuild this table and see if its really fixed? cc @dot2dotseurat
> 
> This table is a critical part of us continuing our V2 migration.

I can rebuild it @bh2smith. We just migrated the locations of our tables and I need to sort out some remaining permissions errors but I will rebuild it asap.

**Comment by dot2dotseurat** [comment]
hi @bh2smith I've just rebuilt the table.

**Comment by bh2smith** [comment]
Thanks very much @dot2dotseurat! I'll check back on this once CI has finished!

**Comment by bh2smith** [comment]
Unfortunately it seems that the Record from September 14 is still missing: I will have a look back and see if there could be any other reason for this:

cf the query with the test - https://dune.com/queries/1389623 

PS - I imagine the later missing records are because of the spells update schedule (or at least I hope so).

**Comment by bh2smith** [comment]
> @bh2smith my hunch would be could be missing from prices but I have no validated. 

Perhaps for the latest records, but I don't believe it's missing prices for sept 14. otherwise we would be seeing a lot more missing than just a single one. Anyway I'll try and look into it soon. Thanks for your help today.

**Comment by bh2smith** [comment]
Ok, I have discovered a new fact that is causing this particular record to be missing:

This Spell Depends on the trades table (which was also missing this records due to this https://github.com/duneanalytics/spellbook/issues/1765). This [missing trades query](https://dune.com/queries/1540086?d=1) shows that there is only one rogue missing entry - all others are from within the last hour (so I chalk this up to scheduling). 

I guess this means the trades table needs to be rebuilt (followed by this one).

[Here is a query](https://dune.com/queries/1539986?+source%28%27prices%27%2C+%27usd%27%29+_t6c1ea=prices.usd&+ref%28%27tokens_ethereum_erc20%27%29+_t6c1ea=tokens_ethereum.erc20&+source%28%27gnosis_protocol_v2_ethereum%27%2C+%27GPv2Settlement_evt_Trade%27%29+_t6c1ea=gnosis_protocol_v2_ethereum.GPv2Settlement_evt_Trade&+source%28%27gnosis_protocol_v2_ethereum%27%2C+%27GPv2Settlement_call_settle%27%29+_t6c1ea=gnosis_protocol_v2_ethereum.GPv2Settlement_call_settle) demonstrating that this trade should be contained in the table.

Note that this is this only "stray" missing record (all others are within the past week). I wonder if this has something to do with the way updates are arranged. Batches depend on Trades how do we ensure that batches are updated after trades or something like that? Or that changes to the trades table trigger updates to the batches?

cc @dot2dotseurat

**Comment by jeff-dude** [comment]
> Batches depend on Trades how do we ensure that batches are updated after trades or something like that? Or that changes to the trades table trigger updates to the batches?

normally in prod, since you use ref() syntax, it would look for trades to refresh first since batches is dependent on it. on the gh action, to save processing time, rather than run all dependencies, we create views of the prod version. that means trades isn't refreshed in this action. the "hacky" way would be to add a minor tweak to that model to get it to run in the same action, then in theory should have updated data.

**Comment by bh2smith** [comment]
Right but this particular transaction is missing in the production trades table so it needs to be rebuilt anyway. Followed by batches (which depends on it) unless that will be triggered anyway.

**Comment by jeff-dude** [comment]
> Right but this particular transaction is missing in the production trades table so it needs to be rebuilt anyway. Followed by batches (which depends on it) unless that will be triggered anyway.

yes, rebuild in prod will happen regardless, but just curious if we should test it now in the gh action too. let me try something.

**Comment by jeff-dude** [comment]
check out the gh action now -- it looks like tests passed on initial build (with incremental run going now). is this expected? previously, you expected a failure with one row returned to find that sept 14th tx.

**Comment by bh2smith** [comment]
If the trades table has been rebuilt then I would expect the missing tx to be fixed. I'm not sure I understand the question.

**Comment by jeff-dude** [comment]
the gh action built trades & batches. then the tests ran and succeeded. then incremental ran successfully and tests succeeded again. if you're happy with the result, i can merge and prod will rebuild

**Review by jeff-dude — APPROVED** [review]

**Comment by bh2smith** [comment]
Are you sure the rebuild will actually be triggered since these changes don't touch the necessary tables?

**Comment by jeff-dude** [comment]
aren't the two issues trades & batches (i.e. edited models in this PR)?
prod ran full refresh on both:
image

**Comment by bh2smith** [comment]
Oh yes! I see you added a dot 8in trades! Thanks. Also checked it out. Looks good so far only missing records that are very recent (which I assume will get covered in the next update). This is incredible thanks a million!

Cc @gentrexha and @josojo.

**Comment by bh2smith** [comment]
Unfortunately, [this query](https://dune.com/queries/1389623) shows that there are STILL missing records from yesterday.

**Comment by dot2dotseurat** [comment]
I can see when your query was run that there were 126 missing trades. When I forked and reran there were 97. To me, this looks like late-arriving data. 

Your model is set up to keep checking for data up to one week back and merging it in every incremental run (we've scheduled that hourly) so my guess is this will resolve in a day. 

```
    {% if is_incremental() %}
    WHERE block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}
```

Screen Shot 2022-11-08 at 9 44 48 AM

Looking at the pct arrived, looking at this [query](https://dune.com/queries/1548038), it looks like we are at 97.4% for yesterday and 100% for all prior days but only 83.4% for today.  

Screen Shot 2022-11-08 at 9 55 43 AM

What's your use case for the data? Are you using it for programmatic decisions or for monitoring/analysis? I'm trying to understand how serious the implications are for late arriving data.

**Comment by bh2smith** [comment]
We are using this data to reimburse protocol operators for their operational costs on a weekly basis. The payout transaction is compiled (partially) from this table about 7 hours past the the previous day.

**Comment by dot2dotseurat** [comment]
I'm not sure. It doesn't seem like we should be more than an hour off. It looks like that is true for trades which is the only other spellbook ref this model is relying upon. 
```
select date_trunc('hour', evt_block_time), count(et.evt_tx_hash), count(t.tx_hash), count(t.tx_hash)/count(et.evt_tx_hash)::float
from gnosis_protocol_v2_ethereum.GPv2Settlement_evt_Trade et
left join cow_protocol_ethereum.trades t 
on et.evt_tx_hash = t.tx_hash
group by 1
order by 1 desc
``` 
Screen Shot 2022-11-08 at 11 18 56 AM

Btw I do see that the spell does rely on one hard coded table that should be a ref (cow_protocol_ethereum.solvers). But that's a view that only uses upstream tables so I don't think the latency could be introduced there.

**Comment by dot2dotseurat** [comment]
This has to be something in the query not the way we run the updates because I am rebuilding it locally with my own schema and it hasn't resolved any issues with late data.

**Comment by bh2smith** [comment]
Found the bug and squashed it! Problem and solution described in https://github.com/duneanalytics/spellbook/pull/1930

Thanks @dot2dotseurat for all your assistance here.

**Pull request merged**